clang/tools/libclang/CDependencies.cpp

```diff
@@ -220,14 +220,25 @@ getFileDependencies(CXDependencyScannerWorker W, int argc,
     *error = cxstring::createEmpty();
 
   DependencyScanningWorker *Worker = unwrap(W);
+  SmallString<128> FullPath;
+
+  auto GetFakeSrcFile = [&]() {
+    FullPath = Worker->getModuleName();
+    llvm::sys::fs::make_absolute(WorkingDirectory, FullPath);
+    return FullPath.c_str();
+  };
 
   std::vector<std::string> Compilation;
-  if (StringRef(argv[1]) == "-cc1")
+  if (StringRef(argv[1]) == "-cc1") {
     for (int i = 2; i < argc; ++i)
       Compilation.push_back(argv[i]);
-  else {
+    if (Worker->getModuleName())
+      Compilation.push_back(GetFakeSrcFile());
```

Comment:
We're already doing that in `DependencyScanningWorker.cpp`, no? Is there a reason we need to inject the name of the fake source file here too?

Reply:
I had to add the argument before `CompilerInvocation::CreateFromArgs`, which gets called in `newInvocation`, gets called. Is there a way to avoid adding the argument or other places where the argument can be added instead of adding it here?

Reply:
And the motivation is that the driver/cc1 command-line parsers in always require an input file? If so, we should document it here.

You mean the call to `newInvocation` from `DependencyScanningWorker::computeDependenciesForClangInvocation`? Or in `ToolInvocation::run`?

I can imagine even the call to `createInvocationFromCommandLine` down on line 244 being problematic, what do you think?

Reply:
I meant the call to `newInvocation` in `DependencyScanningWorker::computeDependenciesForClangInvocation`. The fake source file name has to be added before `CompilerInvocation::CreateFromArgs` is called because `-` will be used as the default input If there is no input file in the argument list (see `ParseFrontendArgs` in `CompilerInvocation.cpp`). 

You are right that the fake source file name is needed for `createInvocationFromCommandLine` too. Without the fake source file, clang will error out early in `Driver::BuildActions` with error message `no input files`.

Reply:
Can you please rebase? With [this PR](https://github.com/apple/llvm-project/pull/3168) merged, none of these command-line modifications should be necessary. Doing the work once in `DependencyScanningWorker` should do it.